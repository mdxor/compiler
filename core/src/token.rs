#[cfg(test)]
use serde::Serialize;

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct Token<T> {
  pub span: Span,
  pub value: T,
}

#[derive(Eq, PartialEq, Debug, Clone)]
#[cfg_attr(test, derive(Serialize))]
pub struct Span {
  pub start: usize,
  pub end: usize,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum HeadingLevel {
  H1,
  H2,
  H3,
  H4,
  H5,
  H6,
}

impl HeadingLevel {
  pub fn new(level: usize) -> Option<HeadingLevel> {
    match level {
      1 => Some(HeadingLevel::H1),
      2 => Some(HeadingLevel::H2),
      3 => Some(HeadingLevel::H3),
      4 => Some(HeadingLevel::H4),
      5 => Some(HeadingLevel::H5),
      6 => Some(HeadingLevel::H6),
      _ => None,
    }
  }

  pub fn to_str(&self) -> &str {
    match self {
      HeadingLevel::H1 => "h1",
      HeadingLevel::H2 => "h2",
      HeadingLevel::H3 => "h3",
      HeadingLevel::H4 => "h4",
      HeadingLevel::H5 => "h5",
      HeadingLevel::H6 => "h6",
    }
  }
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct LinkDefinition<'source> {
  pub label: &'source str,
  pub url: &'source str,
  pub title: String,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum Align {
  Left,
  Center,
  Right,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum BlockToken {
  Paragraph {
    raws: Vec<Span>,
  },
  JSX(JSXElement),
  ATXHeading {
    raws: Vec<Span>,
    level: HeadingLevel,
  },
  SetextHeading {
    raws: Vec<Span>,
    level: HeadingLevel,
  },
  IndentedCode(Vec<Span>),
  BlankLine,
  ThematicBreak,
  BlockQuote {
    blocks: Vec<Token<BlockToken>>,
    level: usize,
  },
  FencedCode {
    meta_span: Span,
    code_spans: Vec<Span>,
  },
  List {
    ch: u8,
    is_tight: bool,
    order_span: Span,
    blocks: Vec<Token<BlockToken>>,
  },
  ListItem {
    indent: usize,
    blocks: Vec<Token<BlockToken>>,
  },
  // LinkDefinition,
  // Table,
  // TableHead,
  // TableCell(&'source str, bool, bool),
  // TableAlignment,
  // TableAlign(Align),
  // TableRow
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum InlineToken {
  TextSegment,
  MaybeLinkStart,
  MaybeEmphasis {
    ch: u8,
    repeat: usize,
    can_open: bool,
    can_close: bool,
  },
  EmphasisStart,
  EmphasisEnd,
  LinkStart {
    url: Span,
    title: Vec<Span>,
  },
  LinkEnd,
  //
  Emphasis(Vec<Token<InlineToken>>),
  Link {
    url: Span,
    title: Vec<Span>,
    text_children: Vec<Token<InlineToken>>,
  },
  Text(Vec<Span>),
  Code(Vec<Span>),
  CodeSegment,
  SoftBreak,
  HardBreak,
  // is email
  AutoLink(bool),
  JSX(JSXElement),
}

#[derive(Eq, PartialEq, Debug)]
pub struct AST<T> {
  pub span: Span,
  pub children: Vec<T>,
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum ContainerBlock {
  BlockQuote(usize),
  List(u8),
  ListItem(usize),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSToken {
  Keyword(Span),
  Punctuator(Span),
  String(Span),
  Template(Span),
  Identifier(Span),
  Number(Span),
  Text(Span),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSXAttr {
  Spread(Vec<JSXExpressionSegment>),
  KeyLiteralValue {
    key: Span,
    value: Span,
  },
  KeyValue {
    key: Span,
    value: Vec<JSXExpressionSegment>,
  },
  KeyTrueValue {
    key: Span,
  },
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSXExpressionSegment {
  JS(Span),
  Element(JSXElement),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub enum JSX {
  Element(JSXElement),
  Text(Span),
  Expression(Vec<JSXExpressionSegment>),
}

#[derive(Eq, PartialEq, Debug)]
#[cfg_attr(test, derive(Serialize))]
pub struct JSXElement {
  pub tag: String,
  pub attributes: Vec<JSXAttr>,
  pub children: Vec<JSX>,
}
