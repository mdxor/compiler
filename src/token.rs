#[derive(Debug, PartialEq)]
pub enum Token<'a> {
  Heading1,
  Heading2,
  Heading3,
  Heading4,
  Heading5,
  Heading6,
  Text(&'a str),
  InlineCode,
}

pub fn get_token(token: &str) -> Token {
  match &*token {
    "#" => Token::Heading1,
    "##" => Token::Heading2,
    "###" => Token::Heading3,
    "####" => Token::Heading4,
    "#####" => Token::Heading5,
    "######" => Token::Heading6,
    _ => Token::Text(token),
  }
}
