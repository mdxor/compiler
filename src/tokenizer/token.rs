#[cfg(test)]
use serde::Serialize;

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct TokenResult<'source> {
  pub size: usize,
  pub token: Token<'source>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub enum Token<'source> {
  ATXHeading(ATXHeading<'source>),
  ThematicBreak,
  SetextHeading(SetextHeading<'source>),
  IndentedCode(IndentedCode<'source>),
  FencedCode(FencedCode<'source>),
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct ATXHeading<'source> {
  pub level: usize,
  pub raw_inlines: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct SetextHeading<'source> {
  pub level: usize,
  pub raw: &'source str,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct IndentedCode<'source> {
  pub codes: Vec<&'source str>,
}

#[derive(Debug, PartialEq)]
#[cfg_attr(test, derive(Serialize))]
pub struct FencedCode<'source> {
  pub code: &'source str,
  pub meta_string: &'source str,
  pub language: &'source str,
}
