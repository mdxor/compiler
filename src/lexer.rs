use crate::jsx;
use crate::rule::Rule;
use crate::token;
use lazy_static::lazy_static;
use std::mem;

lazy_static! {
  static ref RULE: Rule = Rule::new();
}

pub struct Lexer<'a> {
  source: &'a str,
  offset: usize,
  inlines_source: &'a str,
  pub blocks: Vec<token::Block<'a>>,
}

impl<'a> Lexer<'a> {
  pub fn new(source: &'a str) -> Self {
    Lexer {
      source,
      offset: 0,
      inlines_source: "",
      blocks: vec![],
    }
  }

  fn cur(&mut self) -> &'a str {
    &self.source[self.offset..]
  }

  fn skip_whitespace(&mut self) -> usize {
    if let Some(caps) = RULE.whitespace.captures(self.cur()) {
      let size = caps.get(0).unwrap().as_str().len();
      self.move_by(size);
      size
    } else {
      0
    }
  }

  fn move_by(&mut self, size: usize) -> &'a str {
    let result = &self.source[self.offset..self.offset + size];
    self.offset += size;
    result
  }

  fn move_inlines_by(&mut self, size: usize) -> &'a str {
    let result = &self.inlines_source[..size];
    self.inlines_source = &self.inlines_source[size..];
    result
  }

  pub fn tokenize(&mut self) -> Result<token::AST<'a>, &'a str> {
    loop {
      if self.cur().is_empty() {
        if !self.inlines_source.is_empty() {
          let block = self.scan_paragraph()?;
          self.blocks.push(block);
        }
        break;
      } else {
        self.scan_block()?;
      }
    }
    let blocks = mem::take(&mut self.blocks);
    Ok(token::AST { blocks })
  }

  fn scan_block(&mut self) -> Result<(), &'a str> {
    let mut block_option: Option<token::Block<'a>> = None;
    let _inlines_source = self.inlines_source;
    if let Some(block) = self.scan_blank_line() {
      block_option = Some(block);
    } else if let Some(block) = self.scan_single_line_code() {
      block_option = Some(block);
    } else {
      // <=3 whitespace
      self.skip_whitespace();
      if let Some(block) = self.scan_empty_atx_heading() {
        block_option = Some(block);
      } else if let Some(block) = self.scan_heading()? {
        block_option = Some(block);
      } else if let Some(block) = self.scan_fenced_code() {
        block_option = Some(block);
      } else if let Some(block) = self.scan_thematic_break() {
        block_option = Some(block);
      }
    }

    if let Some(block) = block_option {
      if !_inlines_source.is_empty() {
        self.inlines_source = _inlines_source;
        let p = self.scan_paragraph()?;
        self.blocks.push(p);
      }
      self.blocks.push(block);
    } else {
      if let Some(caps) = RULE.line.captures(self.cur()) {
        let size = caps.get(0).unwrap().as_str().len();
        self.move_by(size);
        self.inlines_source =
          &self.source[self.offset - size - self.inlines_source.len()..self.offset];
      }
    }
    Ok(())
  }

  fn scan_blank_line(&mut self) -> Option<token::Block<'a>> {
    if let Some(caps) = RULE.blank_line.captures(self.cur()) {
      let size = caps.get(0).unwrap().as_str().len();
      self.move_by(size);
      Some(token::Block::Leaf(token::LeafBlock::BlankLine))
    } else {
      None
    }
  }

  fn scan_thematic_break(&mut self) -> Option<token::Block<'a>> {
    if let Some(caps) = RULE.thematic_break.captures(self.cur()) {
      let size = caps.get(0).unwrap().as_str().len();
      self.move_by(size);
      Some(token::Block::Leaf(token::LeafBlock::ThematicBreak))
    } else {
      None
    }
  }

  fn scan_paragraph(&mut self) -> Result<token::Block<'a>, &'a str> {
    if self.inlines_source.ends_with("\n") {
      self.inlines_source = &self.inlines_source[..self.inlines_source.len() - 1];
    }
    let inlines = self.scan_inlines()?;

    Ok(token::Block::Leaf(token::LeafBlock::Paragraph(
      token::Paragraph { inlines },
    )))
  }

  fn scan_empty_atx_heading(&mut self) -> Option<token::Block<'a>> {
    if let Some(caps) = RULE.empty_atx_heading.captures(self.cur()) {
      let size = caps.get(1).unwrap().as_str().len();
      self.move_by(caps.get(0).unwrap().as_str().len());
      Some(token::Block::Leaf(token::LeafBlock::Heading(
        token::Heading {
          level: size as u8,
          inlines: vec![],
        },
      )))
    } else {
      None
    }
  }

  fn scan_heading(&mut self) -> Result<Option<token::Block<'a>>, &'a str> {
    if let Some(caps) = RULE.atx_heading.captures(self.cur()) {
      let size = caps.get(1).unwrap().as_str().len();
      let level = size as u8;
      self.move_by(size + 1);
      self.skip_whitespace();

      if let Some(caps) = RULE.line.captures(self.cur()) {
        let size = caps.get(0).unwrap().as_str().len();
        let line_size = caps.get(1).map_or("", |v| v.as_str()).len();
        self.inlines_source = &self.source[self.offset..self.offset + line_size];
        if let Some(caps) = RULE.closing_atx_heading.captures(self.inlines_source) {
          self.inlines_source =
            &self.inlines_source[..line_size - caps.get(0).unwrap().as_str().len()];
        }

        self.move_by(size);
        let inlines = self.scan_inlines()?;
        let heading = token::Heading { level, inlines };
        return Ok(Some(token::Block::Leaf(token::LeafBlock::Heading(heading))));
      }
      return Ok(Some(token::Block::Leaf(token::LeafBlock::Heading(
        token::Heading {
          level: size as u8,
          inlines: vec![],
        },
      ))));
    }
    Ok(None)
  }

  fn scan_inlines(&mut self) -> Result<Vec<token::Inline<'a>>, &'a str> {
    let mut inlines = vec![];
    loop {
      if self.inlines_source.is_empty() {
        return Ok(inlines);
      } else if self.inlines_source.starts_with("`") {
      } else {
        inlines.append(&mut self.scan_inline_text());
      }
    }
  }

  fn scan_inline_text(&mut self) -> Vec<token::Inline<'a>> {
    let mut chars = self.inlines_source.chars();
    let mut text_size = 0;
    let mut texts = vec![];
    let mut escaped = false;
    loop {
      if let Some(char) = chars.next() {
        if !escaped {
          match char {
            '`' => {
              texts.push(token::Inline::Text(self.move_inlines_by(text_size)));
              break;
            }
            '\\' => {
              escaped = !escaped;
            }
            _ => {
              text_size += char.len_utf8();
            }
          }
        } else {
          match char {
            '\\' => {
              if text_size > 0 {
                texts.push(token::Inline::Text(self.move_inlines_by(text_size)));
                text_size = 0;
              }
              texts.push(token::Inline::Text(r"\"));
            }
            '`' | '#' => {
              if text_size > 0 {
                texts.push(token::Inline::Text(self.move_inlines_by(text_size)));
                text_size = 0;
              }
              self.move_inlines_by(1);
              texts.push(token::Inline::Text(self.move_inlines_by(1)));
            }
            _ => {
              text_size += char.len_utf8() + 1;
            }
          }
          escaped = false;
        }
      } else {
        if escaped {
          text_size += 1;
        }
        texts.push(token::Inline::Text(self.move_inlines_by(text_size)));
        break;
      }
    }
    texts
  }

  fn scan_single_line_by_end_char(&mut self, end_char: char) -> &'a str {
    let mut size = 0;
    let mut chars = self.cur().chars();
    loop {
      if let Some(char) = chars.next() {
        if char == end_char || char == '\n' {
          break;
        } else {
          size += char.len_utf8();
        }
      } else {
        return self.move_by(size);
      }
    }
    self.move_by(size + 1)
  }

  fn scan_single_line_code(&mut self) -> Option<token::Block<'a>> {
    if RULE.indented_code.is_match(self.cur()) {
      self.move_by(4);
      let code = self.scan_single_line_by_end_char('\n');
      Some(token::Block::Leaf(token::LeafBlock::IndentedCode(code)))
    } else {
      None
    }
  }

  fn scan_fenced_code(&mut self) -> Option<token::Block<'a>> {
    if RULE.fenced_code.is_match(self.cur()) {
      self.move_by(3);
      let mut language = "";
      let mut metastring = "";
      let mut code = "";
      self.skip_whitespace();
      let mut code_info = self.scan_single_line_by_end_char('\n');
      if code_info.ends_with("\n") {
        code_info = &code_info[..code_info.len() - 1];
      }
      if !code_info.is_empty() {
        if let Some(index) = code_info.find(" ") {
          language = &code_info[0..index].trim();
          metastring = &code_info[index..].trim_start();
        } else {
          language = code_info.trim();
        }
      }
      if let Some(captures) = RULE.fenced_code_end.captures(self.cur()) {
        let end_token = captures.get(0).unwrap().as_str();
        let code_size = self.cur().find(end_token).unwrap();
        code = self.move_by(code_size);
        self.move_by(end_token.len());
      } else {
        code = self.move_by(self.source.len() - self.offset);
      }
      Some(token::Block::Leaf(token::LeafBlock::FencedCode(
        token::FencedCode {
          code,
          language,
          metastring,
        },
      )))
    } else {
      None
    }
  }

  fn scan_jsx(&mut self, is_inline: bool) -> Result<jsx::JSXNode<'a>, &'a str> {
    let mut jsx_parser = jsx::JSXParser::new(self.cur(), 0, is_inline);
    let jsx_node = jsx_parser.parse();
    self.move_by(jsx_parser.size);
    jsx_node
  }
}
