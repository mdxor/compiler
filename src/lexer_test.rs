use crate::lexer;
use crate::token;
fn tokenizes(cases: Vec<&str>) -> Vec<Result<token::AST, &str>> {
  let mut results = vec![];
  for case in &cases {
    let mut lex = lexer::Lexer::new(case);
    let ast = lex.tokenize();
    results.push(ast)
  }
  results
}

#[test]
fn test_indented_code() {
  let cases = vec!["    abc", "    <div></div>"];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
#[test]
fn test_fenced_code() {
  let cases = vec![
    "```\ncode\n```",
    "```jsx\nlet a = 11;\n```",
    "```jsx meta\nlet a = 11;\n```",
  ];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
#[test]
fn test_atx_heading() {
  let cases = vec![
    // "# foo",
    // "## foo",
    // "### foo",
    // "#### foo",
    // "##### foo",
    // "###### foo",
    // "####### foo",
    // "#5 bolt",
    // "#hashtag",
    // r"\## foo",
    // // TODO:
    // // r"# foo *bar* \*baz\*"
    // "#                  foo                     ",
    // "### foo",
    // " ## foo",
    // "  # foo",
    // "    # foo",
    // // "foo\n    # bar",
    // "## foo ##\n###   bar    ###",
    // "# foo ##################################\n##### foo ##",
    // "### foo ###     ",
    // "### foo ### b",
    // "# foo#",
    // "****\n## foo\n****"
    "Foo bar\n# baz\nBar foo",
    // "## \n#\n### ###",
  ];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
#[test]
fn test_thematic_break() {
  let cases = vec![
    "***\n---\n___",
    "+++",
    "===",
    "--\n**\n__",
    "***\n ***\n   ***",
    "    ***",
    // "Foo\n    ***",
    "_____________________________________",
    "- - -",
    " **  * ** * ** * **",
    "-     -      -      -",
    "- - - -    ",
    "_ _ _ _ a\na------\n---a---",
    // " *-*",
    // "- foo\n***\n- bar"
    "Foo\n***\nbar",
    // "Foo\n---\nbar",
    // "* Foo\n* * *\n* Bar",
    // "- Foo\n- * * *",
  ];
  insta::assert_yaml_snapshot!(tokenizes(cases));
}
