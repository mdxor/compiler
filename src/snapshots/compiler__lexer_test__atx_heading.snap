---
source: src/lexer_test.rs
expression: tokenizes(cases)

---
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 1
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 2
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 3
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 4
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 5
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 6
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Paragraph:
            inlines:
              - Text: "####### foo"
- Ok:
    blocks:
      - Leaf:
          Paragraph:
            inlines:
              - Text: "#5 bolt"
- Ok:
    blocks:
      - Leaf:
          Paragraph:
            inlines:
              - Text: "#hashtag"
- Ok:
    blocks:
      - Leaf:
          Paragraph:
            inlines:
              - Text: "#"
              - Text: "# foo"
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 1
            inlines:
              - Text: "foo                     "
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 3
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 2
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 1
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          IndentedCode: "# foo"
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 2
            inlines:
              - Text: foo
      - Leaf:
          Heading:
            level: 3
            inlines:
              - Text: "bar   "
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 1
            inlines:
              - Text: foo
      - Leaf:
          Heading:
            level: 5
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 3
            inlines:
              - Text: foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 3
            inlines:
              - Text: "foo ### b"
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 1
            inlines:
              - Text: "foo#"
- Ok:
    blocks:
      - Leaf:
          Paragraph:
            inlines:
              - Text: Foo bar
      - Leaf:
          Heading:
            level: 1
            inlines:
              - Text: baz
      - Leaf:
          Paragraph:
            inlines:
              - Text: Bar foo
- Ok:
    blocks:
      - Leaf:
          Heading:
            level: 2
            inlines: []
      - Leaf:
          Heading:
            level: 1
            inlines: []
      - Leaf:
          Heading:
            level: 3
            inlines: []

